src/storage-client/src/client.rs

```diff
@@ -549,6 +599,24 @@ where
                     }
                 }
             }
+            StorageCommand::AlterSources(alter_ingestions) => {
+                for ingestion in alter_ingestions {
+                    for export_id in ingestion.description.subsource_ids() {
+                        if self.uppers.contains_key(&export_id) {
+                            continue;
+                        }
+
+                        let mut frontier = MutableAntichain::new();
+                        // TODO(guswynn): cluster-unification: fix this dangerous use of `as`, by
+                        // merging the types that compute and storage use.
+                        #[allow(clippy::as_conversions)]
+                        frontier.update_iter(iter::once((T::minimum(), self.parts as i64)));
+                        let part_frontiers =
+                            vec![Some(Antichain::from_elem(T::minimum())); self.parts];
+                        self.uppers.insert(export_id, (frontier, part_frontiers));
```

Comment:
Could it be that we have to remove uppers for exports that now don't exist anymore? Same question for the rehydration shim.

Reply:
There's an impedance mismatch between creating and dropping subsources where the lower-level clients don't get uppers for subsources until their primary source/ingestion gets created, but they get dropped as if they were primary sources, i.e. the coordinator is responsible for dropping them, where the `AllowCompaction` machinery removes them from the lower-level clients.

Reply:
let's chat about this before you merge. I'm not sure yet that I understand completely 😅

Reply:
We chatted about this over zoom and came to a shared understanding of what's going on here.